File: src/library/java/gg/generations/rarecandy/pokeutils/IVariantDetails.java
package gg.generations.rarecandy.pokeutils;

import com.google.gson.*;
import gg.generations.rarecandy.pokeutils.util.JsonUtils;
import gg.generations.rarecandy.renderer.animation.ITransformSet;

import java.util.Objects;

public interface IVariantDetails {
    String material();
    String effect();
    Boolean paradox();
    Boolean hide();
    ITransformSet transform();

    static JsonElement serialize(IVariantDetails variantDetails) {
        var obj = new JsonObject();
        JsonUtils.putIf(Objects::nonNull, obj, "material", variantDetails.material(), JsonPrimitive::new);
        JsonUtils.putIf(Objects::nonNull, obj, "hide", variantDetails.hide(), JsonPrimitive::new);
        JsonUtils.putIf(Objects::nonNull, obj, "paradox", variantDetails.paradox(), JsonPrimitive::new);
        JsonUtils.putIf(transform -> transform != null && !transform.isUnit(), obj, "transform", variantDetails.transform(), ITransformSet::serialize);
        return obj;
    }

    static IVariantDetails deserialize(JsonElement jsonElement) {
        var obj = jsonElement.getAsJsonObject();
        var material = JsonUtils.extractIfPresent(obj, "material", null, JsonElement::getAsString);
        var effect = JsonUtils.extractIfPresent(obj, "effect", null, JsonElement::getAsString);
        var paradox = JsonUtils.extractIfPresent(obj, "paradox", null, JsonElement::getAsBoolean);
        var hide = JsonUtils.extractIfPresent(obj, "hide", null, JsonElement::getAsBoolean);
        var offset = JsonUtils.extractIfPresent(obj, "transform", null, ITransformSet::deserialize);
        return IModelConfig.Factory.ACTIVE_FACTORY.createVariantDetails(material, effect, paradox, hide, offset);
    }
}

File: src/library/java/gg/generations/rarecandy/pokeutils/IVariantDetails.kt
package gg.generations.rarecandy.pokeutils

import gg.generations.rarecandy.codec.BOOL
import gg.generations.rarecandy.codec.STRING
import gg.generations.rarecandy.codec.codec
import gg.generations.rarecandy.codec.obj
import gg.generations.rarecandy.codec.putIfNotNull
import gg.generations.rarecandy.codec.readOrNull
import gg.generations.rarecandy.renderer.animation.ITransformSet

data class IVariantDetails(
    var material: String?,
    var effect: String?,
    var paradox: Boolean?,
    var hide: Boolean?,
    var transform: ITransformSet?) {

    constructor(): this(null, null, null, null, null)

    companion object {
        val CODEC = codec({ set ->
            obj().also {
                it.putIfNotNull("material", STRING, set.material)
                it.putIfNotNull("effect", STRING, set.effect)
                it.putIfNotNull("paradox", BOOL, set.paradox)
                it.putIfNotNull("hide", BOOL, set.hide)
                it.putIfNotNull("transform", ITransformSet.CODEC, set.transform)
            }
        }, {
            it.obj().let {
                IVariantDetails(
                    it.readOrNull("material", STRING),
                    it.readOrNull("effect", STRING),
                    it.readOrNull("paradox", BOOL),
                    it.readOrNull("hide", BOOL),
                    it.readOrNull("transform", ITransformSet.CODEC)
                )
            }
        })
    }
}
